### PR TITLE
feat: branch encoding allows 256bits separator

### PR DESCRIPTION
**Pull request by gabriele-0201**

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/389?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#389** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/389?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#385** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/385?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#384** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/384?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#383** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/383?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/beatree/ops/mod.rs [hunk 1]
```diff
@@ -59,24 +60,13 @@ pub fn lookup(
...
-
-    let post_key = &key.view_bits::<Msb0>()[prefix.len()..];
-
     let mut low = 0;
     let mut high = branch.n() as usize;
 
     while low < high {
         let mid = low + (high - low) / 2;
-        if post_key < branch.separator(mid) {
+        let s = reconstruct_key(prefix, branch.separator(mid));
```
Comment (rphmeier) [inline nomt/src/beatree/ops/mod.rs:68]: remember: we _don't_ want to pessimize the typical case. this pessimizes the typical case by doing more work.

File: nomt/src/beatree/branch/node.rs [hunk 2]
```diff
@@ -24,6 +24,13 @@ use crate::beatree::Key;
 //
 // node_pointers: LNPN or BNID[n]
 // ```
+//
+// There is also the possibility of specifying one or more separators
+// that do not share the prefix specified at the beginning of the node.
+// This is the case only if the separator length extracted by the cells is 256.
```
Comment (rphmeier) [inline nomt/src/beatree/branch/node.rs:30]: this is incorrect.

File: nomt/src/beatree/ops/mod.rs [hunk 3]
```diff
@@ -59,24 +60,13 @@ pub fn lookup(
 /// node pointer whose separator is less than or equal to the given key.
 fn search_branch(branch: &branch::BranchNode, key: Key) -> Option<(usize, PageNumber)> {
     let prefix = branch.prefix();
-
-    match key.view_bits::<Msb0>()[..prefix.len()].cmp(prefix) {
-        Ordering::Equal => {}
-        Ordering::Less => return None,
-        Ordering::Greater => {
```
Comment (rphmeier) [inline nomt/src/beatree/ops/mod.rs:66]: actually, the only thing that needs to change here is that this branch needs a match guard `if k == n`. then just have a small branch for the comparisons in the binary search.

**Pull request closed without merging**